src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,221 @@
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> None:
+        """Update the buffer with a new value for the given timestamp.
+
+        Args:
+            timestamp: Timestamp of the new value
+            value: value to add
+            missing: if true, the given timestamp will be recorded as missing.
+                The value will still be written.
+
+            Returns:
+                Nothing.
+        """
+        # Update timestamps
+        self._datetime_newest = max(self._datetime_newest, timestamp)
+        self._datetime_oldest = min(self._datetime_oldest, timestamp)
+
+        if self._datetime_oldest < self._datetime_newest - timedelta(
+            seconds=len(self._buffer) * self._sample_rate
+        ):
+            self._datetime_oldest = self._datetime_newest - timedelta(
+                len(self._buffer) * self._sample_rate
+            )
+
+        # Update data
+        insert_index = self.datetime_to_index(timestamp)
+
+        self._buffer[insert_index] = value
+
+        # Update list of missing windows
+        #
+        # We always append to the last pending window.
+        # A window is pending when end is None
+        if missing:
+            # Create new if no pending window
+            if (
+                len(self._missing_windows) == 0
+                or self._missing_windows[-1].end is not None
+            ):
+                self._missing_windows.append({"start": timestamp, "end": None})
+        elif len(self._missing_windows) > 0:
+            # Finalize a pending window
+            if self._missing_windows[-1].end is None:
+                self._missing_windows[-1].end = timestamp
+
+        # Delete out-to-date windows
+        if len(self._missing_windows) > 0 and self._missing_windows[0].end is not None:
+            if self._missing_windows[0].end <= self._datetime_oldest:
+                self._missing_windows = self._missing_windows[1:]
+
+    def datetime_to_index(self, timestamp: datetime) -> int:
+        """Convert the given timestamp to an index.
+
+        Throws an index error when the timestamp is not found within this
+        buffer.
+
+        Args:
+            timestamp: Timestamp to convert.
+
+        Raises:
+            IndexError: when requesting a timestamp outside the range this container holds
+
+        Returns:
+            index where the value for the given timestamp can be found.
+        """
+        if self._datetime_newest < timestamp or timestamp < self._datetime_oldest:
+            raise IndexError(
+                f"Requested timestamp {timestamp} is is "
+                f"outside the range [{self._datetime_oldest} - {self._datetime_newest}]"
+            )
+
+        return self._wrap(int(abs((self._window_start - timestamp).total_seconds())))
```

Comment:
Shouldn't this be divided by the sample rate?

Reply:
Indeed, my test was all set to 1 and that's why I didn't catch it. Nice catch!